### PR TITLE
Datastore Garbage Collection

### LINKED ISSUE #64: Datastore Garbage Collection
We currently have no mechanism for deleting objects from the datastore; this is partly by design, as we want the datastore to behave as an immutable append-only store, and partly by implementation constraints as tracking objects in the datastore is expensive (storage cost, performance effects on the ingestion critical path).

Nonetheless, we do need a mechanism for pruning the local datastore, for example when an operator decides to unmerge some large dataset.

Here are two possible approaches:
- Directly track all object deletions and at GC time iterate through the statements in StatementDB to find unreferenced objects and delete them. The downside is that we need to track object deletions, which has storage and complexity cost. It's also not clear if there will be any space saving in the datastore, at least not without compaction.
- Use a semi-space garbage collection. This is the superior approach in terms of implementation simplicity and also most effective pruning of the datastore. With the semi space approach, at GC we create a new rocks db datastore open for bulk load; We then iterate through statements in the db, and copy all referenced objects from the current datastore the new space. When we are done, we compact the new datastore and delete the old one. The downside is that it requires free space in the file system for the copy, which may not be feasible for capped nodes. This may be remedied by deleting copied objects from the old datastore, and compacting it at regular intervals, although this adds complexity and has the same caveats about deletions effecting actual datastore size.

### PR DESCRIPTION
**Pull request by vyzo**
Closes #64 
Implements datastore garbage collection with direct iteration.

Implementation Notes:
- datastore iterators require flushing the db to work reliably
- a temporary db is used for storing the live object index
- `/data/compact` api endpoint for manual compaction; gc triggers it if it deletes any objects, but it's also useful on its own right
- `/data/sync` api endpoint: the WAL seems to be using some dark space in linux/ext4 (not visible in ls, but visible in du) following merges, but it seems to be bounded by the log size. Flushing the datastore through the endpoint immediately reclaims this space,
- `/data/keys` api endpoint which dumps the keys for all objects in the datastore.

Example:
```
$ du --si /mnt/ssd2/data/
193k    /mnt/ssd2/data/
$ mcclient status online
status set to online
$ time mcclient merge QmeiY2eHMwK92Zt6X4kUUC3MsjMmVb2VnGZ17DhnhRPCEQ "SELECT * FROM images.dpla LIMIT 1000000"
merged 1000000 statements and 1000001 objects

real    6m48.171s
user    0m0.345s
sys     0m0.033s
$ du --si /mnt/ssd2/data/
867M    /mnt/ssd2/data/
$ time mcclient delete "DELETE FROM images.dpla"  
Deleted 1000000 statements

real    1m2.060s
user    0m0.332s
sys     0m0.041s
$ mcclient status offline
status set to offline
$ time curl http://127.0.0.1:9002/data/gc
1000001

real    0m57.923s
user    0m0.000s
sys     0m0.008s
$ du --si /mnt/ssd2/data/
193k    /mnt/ssd2/data/
```

### CODE REVIEW

File: mcnode/ds.go [hunk 1]
```diff
@@ -105,6 +108,42 @@ func (ds *RocksDS) Delete(key Key) error {
 	return ds.db.Delete(ds.wo, key[2:])
 }
 
+func (ds *RocksDS) Sync() error {
+	return ds.db.Flush(ds.fo)
+}
+
+func (ds *RocksDS) IterKeys(ctx context.Context) (<-chan Key, error) {
```
Comment (parkan) [inline mcnode/ds.go]: Sounds like at this point we are 90% of the way to implementing ipfs/go-datastore

File: mcnode/gc.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+		return err
+	}
+
+	insertKey, err := db.Prepare("INSERT INTO Refs VALUES (?)")
+	if err != nil {
+		return err
+	}
+	gc.insertKey = insertKey
+
+	countKeys, err := db.Prepare("SELECT COUNT(1) FROM Refs WHERE key = ?")
```
Comment (parkan) [inline mcnode/gc.go]: Given that you're checking for existence (> 0) this should have a `LIMIT 1` to stop scanning after finding a match. I also wonder if `SELECT EXISTS(SELECT 1 FROM Refs WHERE key = ?)` is gonna get optimized to some nice index scan, maybe ask `EXPLAIN`?

Comment (vyzo) [inline mcnode/gc.go]: It's a primary key scan, so I would expect it to stop scanning.

File: mcnode/ds.go (hunk 1 above)
Comment (vyzo) [inline mcnode/ds.go]: Our datastore is more specialized (only uses hashes for keys).

File: mcnode/gc.go (hunk 2 above)
Comment (parkan) [inline mcnode/gc.go]: I guess I'm overthinking this :)

File: mcnode/ds.go (hunk 1 above)
Comment (vyzo) [inline mcnode/ds.go]: It's probably ok, but the key=hash invariant is more widespread than just Put/PutBatch; the key type is a typedef for a multihash, and there are several places where the keys are decoded as mhashes.
Having said that, we could in theory extend to a more generalized interface, but I am not sure we'll gain anything from it.
Also, there is the issue of tracking references for arbitrary object, garbage collection would just reap whatever data we put in.

**Pull request merged**